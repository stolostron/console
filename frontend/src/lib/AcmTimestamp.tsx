/* Copyright Contributors to the Open Cluster Management project */

import { css } from '@emotion/css'
import { SimpleTimestamp } from './SimpleTimestamp'
import { PluginContext } from './PluginContext'
import { useContext } from 'react'
import { TimestampProps } from '@openshift-console/dynamic-plugin-sdk'

type AcmTimestampProps = TimestampProps & {
  showIcon?: boolean
}

const AcmTimestamp: React.FC<AcmTimestampProps> = ({
  timestamp,
  simple,
  omitSuffix,
  className = '',
  showIcon = false,
}) => {
  const {
    ocpApi: { Timestamp },
  } = useContext(PluginContext)

  return Timestamp ? (
    <Timestamp
      timestamp={timestamp}
      simple={simple}
      omitSuffix={omitSuffix}
      className={`${className}${
        !showIcon
          ? ` ${css({
              '.co-icon-and-text__icon': {
                display: 'none',
              },
            })}`
          : ''
      }`}
    />
  ) : (
    <SimpleTimestamp timestamp={timestamp} />
  )
}

export default AcmTimestamp
