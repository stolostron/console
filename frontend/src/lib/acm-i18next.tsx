/* Copyright Contributors to the Open Cluster Management project */
// eslint-disable-next-line @typescript-eslint/no-restricted-imports
import { useTranslation as useReactI18NextTranslation, Trans as ReactTrans } from 'react-i18next'

export function useTranslation() {
  return useReactI18NextTranslation(process.env.TRANSLATION_NAMESPACE)
}

export function Trans(props: any) {
  const { t } = useTranslation()
  return (
    <ReactTrans {...props} t={t}>
      {props.children}
    </ReactTrans>
  )
}
