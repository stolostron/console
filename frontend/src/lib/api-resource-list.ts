/* Copyright Contributors to the Open Cluster Management project */

import { getBackendUrl, getRequest } from '../resources'

const versionUrl = '/apiPaths'

export interface APIResourceNames {
  [kind: string]: APIResourceMeta
}

export interface APIResourceMeta {
  [kind: string]: { pluralName: string }
}

export function getApiPaths() {
  const url = getBackendUrl() + versionUrl
  return getRequest<APIResourceNames>(url)
}
