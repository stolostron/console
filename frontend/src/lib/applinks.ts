/* Copyright Contributors to the Open Cluster Management project */

import { getBackendUrl, getRequest } from '../resources'

export const applinksUrl = '/console-links'

export type IAppLinksResult = {
    data: Record<string, [IAppSwitcherData]>
}

export interface IAppSwitcherData {
    name: string
    url: string
    icon: string
}

export function getApplinks() {
    const url = getBackendUrl() + applinksUrl
    return getRequest<IAppLinksResult>(url)
}
