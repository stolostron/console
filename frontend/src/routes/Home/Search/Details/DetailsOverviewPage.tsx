/* Copyright Contributors to the Open Cluster Management project */
import {
  Button,
  DescriptionList,
  DescriptionListDescription,
  DescriptionListGroup,
  DescriptionListTerm,
  Divider,
  Flex,
  FlexItem,
  Label,
  LabelGroup,
  PageSection,
  Stack,
  Text,
  Tooltip,
} from '@patternfly/react-core'
import { GlobeAmericasIcon, PencilAltIcon, SearchIcon } from '@patternfly/react-icons'
import _ from 'lodash'
import { Fragment, useEffect, useMemo, useState } from 'react'
import { generatePath, Link, useHistory } from 'react-router-dom'
import { findResourceFieldLineNumber } from '../../../../components/YamlEditor'
import { useTranslation } from '../../../../lib/acm-i18next'
import { canUser } from '../../../../lib/rbac-util'
import { NavigationPath } from '../../../../NavigationPath'
import { IResource, OwnerReference } from '../../../../resources'
import { AcmAlert, AcmLoadingPage, AcmTable } from '../../../../ui-components'

export function ResourceSearchLink(props: {
  cluster: string
  apiversion: string
  kind: string
  name: string
  namespace?: string
}) {
  const { cluster, kind, name, namespace, apiversion } = props
  let searchParams = `?filters={"textsearch":"cluster%3A${cluster}%20kind%3A${kind}%20name%3A${name}`
  if (namespace) {
    searchParams = `${searchParams}%20namespace%3A${namespace}`
  }
  if (apiversion.includes('/')) {
    searchParams = `${searchParams}%20apigroup%3A${apiversion.split('/')[0]}%20apiversion%3A${apiversion.split('/')[1]}`
  } else {
    searchParams = `${searchParams}%20apiversion%3A${apiversion}`
  }
  return (
    <Link
      to={{
        pathname: NavigationPath.search,
        search: searchParams,
      }}
    >
      {name}
    </Link>
  )
}

export function LablesGroup(props: { labels: Record<string, string> }) {
  const { t } = useTranslation()
  const { labels } = props
  if (Object.keys(labels).length === 0) {
    return (
      <LabelGroup
        style={{
          color: 'var(--pf-global--Color--200)',
          fontSize: '14px',
        }}
      >
        {t('No labels')}
      </LabelGroup>
    )
  }
  return (
    <LabelGroup>
      {Object.keys(labels).map((key) => (
        <Label key={key} id={key} color="blue">
          {key}
          {typeof labels[key] === 'string' && labels[key].trim() !== '' && `=${labels[key]}`}
        </Label>
      ))}
    </LabelGroup>
  )
}

export function getDate(timestamp?: string) {
  if (!timestamp) {
    return '-'
  }
  const mdate = new Date(timestamp)
  const localDateTime = new Intl.DateTimeFormat(undefined, {
    month: 'short',
    day: 'numeric',
    year: 'numeric',
    hour: 'numeric',
    minute: 'numeric',
  })
  const utcDateTime = new Intl.DateTimeFormat(undefined, {
    month: 'short',
    day: 'numeric',
    hour: 'numeric',
    minute: 'numeric',
    year: 'numeric',
    timeZone: 'UTC',
    timeZoneName: 'short',
  })
  return (
    <div style={{ display: 'flex', alignItems: 'baseline' }}>
      <GlobeAmericasIcon style={{ marginRight: '5px' }} />
      <Tooltip
        content={[
          <span className="co-nowrap" key="co-timestamp">
            {utcDateTime.format(mdate)}
          </span>,
        ]}
      >
        <span data-test="timestamp">{localDateTime.format(mdate)}</span>
      </Tooltip>
    </div>
  )
}

export function OwnerReferences(props: { ownerReferences?: OwnerReference[]; namespace?: string; cluster: string }) {
  const { ownerReferences, namespace, cluster } = props
  const { t } = useTranslation()
  if (!ownerReferences || ownerReferences.length === 0) {
    return <div>{t('No owners')}</div>
  }
  return (
    <ResourceSearchLink
      cluster={cluster}
      kind={ownerReferences[0].kind}
      name={ownerReferences[0].name}
      namespace={namespace}
      apiversion={ownerReferences[0].apiVersion}
    />
  )
}

interface ResourceCondition {
  type: string
  status: 'True' | 'False' | 'Unknown'
  lastTransitionTime?: string
  reason?: string
  message?: string
}
export function ResourceConditions(props: { conditions: ResourceCondition[] }) {
  const { conditions } = props
  const { t } = useTranslation()

  const cols = useMemo(
    () => [
      {
        header: 'Type',
        cell: 'type',
      },
      {
        header: 'Status',
        cell: 'status',
      },
      {
        header: 'Updated',
        cell: (item: ResourceCondition) => {
          if (!item.lastTransitionTime) {
            return '-'
          }
          return getDate(item.lastTransitionTime)
        },
      },
      {
        header: 'Reason',
        cell: (item: ResourceCondition) => {
          if (!item.reason) {
            return '-'
          }
          return item.reason
        },
      },
      {
        header: 'Message',
        cell: (item: ResourceCondition) => {
          if (!item.message) {
            return '-'
          }
          return item.message
        },
      },
    ],
    []
  )

  return (
    <Fragment>
      <Divider />
      <PageSection variant={'light'}>
        <Text style={{ fontSize: '1.25rem', fontFamily: 'RedHatDisplay' }} component={'h2'}>
          {'Conditions'}
        </Text>
        {conditions.length ? (
          <AcmTable
            key={'resource-conditions'}
            plural={'resource conditions'}
            items={conditions}
            columns={cols}
            keyFn={() => Math.random().toString(36).substring(7)}
            autoHidePagination={true}
          />
        ) : (
          t('No conditions found')
        )}
      </PageSection>
    </Fragment>
  )
}

export default function DetailsOverviewPage(props: {
  cluster: string
  resource: IResource
  loading: boolean
  error: string
}) {
  const { cluster, resource, loading, error } = props
  const { t } = useTranslation()
  const history = useHistory()
  const [canEditResource, setCanEditResource] = useState<boolean>(false)

  const { labelsLineNumber, annotationsLineNumber, tolerationsLineNumber } = useMemo(() => {
    let tolerationsPath = '/spec/tolerations'
    if (resource) {
      if (resource.kind.toLowerCase() !== 'pod') {
        tolerationsPath = '/spec/template/spec/tolerations'
      }
      const labelsLineNumber = findResourceFieldLineNumber(resource, '/metadata/labels')
      const annotationsLineNumber = findResourceFieldLineNumber(resource, '/metadata/annotations')
      const tolerationsLineNumber = findResourceFieldLineNumber(resource, tolerationsPath)
      return { labelsLineNumber, annotationsLineNumber, tolerationsLineNumber }
    }
    return { labelsLineNumber: -1, annotationsLineNumber: -1, tolerationsLineNumber: -1 }
  }, [resource])

  useEffect(() => {
    if (resource) {
      const canDeleteSubscriptionPromise = canUser('update', {
        apiVersion: resource.apiVersion,
        kind: resource.kind,
        metadata: {
          name: resource.metadata?.name ?? '',
          namespace: resource.metadata?.namespace ?? '',
        },
      })
      canDeleteSubscriptionPromise.promise
        .then((result) => setCanEditResource(result.status?.allowed!))
        .catch((err) => console.error(err))
      return () => canDeleteSubscriptionPromise.abort()
    }
  }, [resource])

  const resourceTolerationsCount: number = useMemo(() => {
    if (resource && resource.kind.toLowerCase() === 'pod') {
      const podTolerations = _.get(resource, 'spec.tolerations', {})
      return podTolerations.length
    }
    const resourceTolerations = _.get(resource, 'spec.template.spec.tolerations', {})
    return resourceTolerations.length
  }, [resource])

  const podSelectorLink = useMemo(() => {
    if (resource) {
      const requirements: any = []
      const podSelectors = _.get(resource, 'spec.selector', {})
      const matchLabels =
        !podSelectors.matchLabels && !podSelectors.matchExpressions ? podSelectors : podSelectors.matchLabels

      Object.keys(matchLabels || {})
        .sort()
        .forEach(function (k) {
          requirements.push(`${k}=${matchLabels[k]}`)
        })

      if (requirements.length > 0) {
        const searchParams = encodeURIComponent(`cluster:${cluster} kind:Pod label:${requirements.join(',')}`)
        return (
          <Button
            data-test="pod-selector-btn"
            isInline
            icon={<SearchIcon />}
            onClick={() => {
              history.push(`${NavigationPath.search}?filters={"textsearch":"${searchParams}"}`)
            }}
            variant="link"
          >
            {requirements.join(', ')}
          </Button>
        )
      }
    }
  }, [cluster, resource, history])

  const nodeSelectorLink = useMemo(() => {
    if (resource) {
      const requirements: any = []
      const nodeSelectors = _.get(resource, 'spec.template.spec.nodeSelector', {})
      Object.keys(nodeSelectors || {})
        .sort()
        .forEach(function (k) {
          requirements.push(nodeSelectors[k] !== '' ? `${k}=${nodeSelectors[k]}` : k)
        })

      if (requirements.length > 0) {
        const searchParams = encodeURIComponent(`cluster:${cluster} kind:Node label:${requirements.join(',')}`)
        return (
          <Button
            data-test="node-selector-btn"
            isInline
            icon={<SearchIcon />}
            onClick={() => {
              history.push(`${NavigationPath.search}?filters={"textsearch":"${searchParams}"}`)
            }}
            variant="link"
          >
            {requirements.join(', ')}
          </Button>
        )
      }
    }
  }, [cluster, resource, history])

  if (error) {
    return (
      <PageSection>
        <AcmAlert
          noClose={true}
          variant={'danger'}
          isInline={true}
          title={`${t('Error querying for resource:')} ${name}`}
          subtitle={error}
        />
      </PageSection>
    )
  } else if (loading) {
    return (
      <PageSection>
        <AcmLoadingPage />
      </PageSection>
    )
  }

  if (resource && !loading && !error) {
    return (
      <PageSection>
        <PageSection variant={'light'}>
          <Stack hasGutter>
            <Text style={{ fontSize: '1.25rem', fontFamily: 'RedHatDisplay' }} component={'h2'}>
              {`${resource.kind} details`}
            </Text>
            <DescriptionList
              columnModifier={{
                default: '2Col',
              }}
              style={{ fontSize: '14px' }}
            >
              <DescriptionListGroup>
                <DescriptionListTerm>{'Name'}</DescriptionListTerm>
                <DescriptionListDescription>{resource.metadata?.name}</DescriptionListDescription>
              </DescriptionListGroup>

              {resource.metadata?.namespace && (
                <DescriptionListGroup>
                  <DescriptionListTerm>{'Namespace'}</DescriptionListTerm>
                  <DescriptionListDescription>
                    <Button
                      data-test="namespace-nav-link"
                      type="button"
                      isInline
                      onClick={() => {
                        history.push(
                          `${NavigationPath.resources}?cluster=${cluster}&kind=Namespace&apiversion=v1&name=${resource.metadata?.namespace}`
                        )
                      }}
                      variant="link"
                    >
                      {resource.metadata?.namespace}
                    </Button>
                  </DescriptionListDescription>
                </DescriptionListGroup>
              )}

              <DescriptionListGroup>
                <DescriptionListTerm>{'Cluster'}</DescriptionListTerm>
                <DescriptionListDescription>
                  <Link
                    to={{
                      pathname: generatePath(NavigationPath.clusterOverview, {
                        name: cluster,
                        namespace: cluster,
                      }),
                    }}
                  >
                    {cluster}
                  </Link>
                </DescriptionListDescription>
              </DescriptionListGroup>

              <DescriptionListGroup>
                <Flex justifyContent={{ default: 'justifyContentSpaceBetween' }}>
                  <FlexItem>
                    <DescriptionListTerm>{'Labels'}</DescriptionListTerm>
                  </FlexItem>
                  {canEditResource && (
                    <FlexItem>
                      <Link
                        to={{
                          pathname: NavigationPath.resourceYAML,
                          search: window.location.search,
                          state: { scrollToLine: labelsLineNumber },
                        }}
                      >
                        {t('Edit')}
                        <PencilAltIcon style={{ marginLeft: '.5rem' }} />
                      </Link>
                    </FlexItem>
                  )}
                </Flex>
                <DescriptionListDescription
                  style={{
                    border: '1px solid var(--pf-global--BorderColor--300)',
                    borderRadius: 'var(--pf-c-label-group--m-category--BorderRadius)',
                    padding: '0.25rem',
                  }}
                >
                  <LablesGroup labels={resource.metadata?.labels ?? {}} />
                </DescriptionListDescription>
              </DescriptionListGroup>

              {podSelectorLink && (
                <DescriptionListGroup>
                  <DescriptionListTerm>{'Pod selector'}</DescriptionListTerm>
                  <DescriptionListDescription>{podSelectorLink}</DescriptionListDescription>
                </DescriptionListGroup>
              )}

              {nodeSelectorLink && (
                <DescriptionListGroup>
                  <DescriptionListTerm>{'Node selector'}</DescriptionListTerm>
                  <DescriptionListDescription>{nodeSelectorLink}</DescriptionListDescription>
                </DescriptionListGroup>
              )}

              {resourceTolerationsCount > 0 && (
                <DescriptionListGroup>
                  <DescriptionListTerm>{'Tolerations'}</DescriptionListTerm>
                  <DescriptionListDescription>
                    {canEditResource ? (
                      <Link
                        to={{
                          pathname: NavigationPath.resourceYAML,
                          search: window.location.search,
                          state: { scrollToLine: tolerationsLineNumber },
                        }}
                      >
                        {t('tolerations.count', {
                          count: resourceTolerationsCount,
                        })}
                        <PencilAltIcon style={{ marginLeft: '.5rem' }} />
                      </Link>
                    ) : (
                      t('tolerations.count', {
                        count: resourceTolerationsCount,
                      })
                    )}
                  </DescriptionListDescription>
                </DescriptionListGroup>
              )}

              <DescriptionListGroup>
                <DescriptionListTerm>{'Annotations'}</DescriptionListTerm>
                <DescriptionListDescription>
                  {canEditResource ? (
                    <Link
                      to={{
                        pathname: NavigationPath.resourceYAML,
                        search: window.location.search,
                        state: { scrollToLine: annotationsLineNumber },
                      }}
                    >
                      {t('annotations.count', {
                        count: Object.keys(resource.metadata?.annotations ?? {}).length,
                      })}
                      <PencilAltIcon style={{ marginLeft: '.5rem' }} />
                    </Link>
                  ) : (
                    t('annotations.count', {
                      count: Object.keys(resource.metadata?.annotations ?? {}).length,
                    })
                  )}
                </DescriptionListDescription>
              </DescriptionListGroup>

              <DescriptionListGroup>
                <DescriptionListTerm>{'Created at'}</DescriptionListTerm>
                <DescriptionListDescription>{getDate(resource.metadata?.creationTimestamp)}</DescriptionListDescription>
              </DescriptionListGroup>

              <DescriptionListGroup>
                <DescriptionListTerm>{'Owner'}</DescriptionListTerm>
                <DescriptionListDescription>
                  <OwnerReferences
                    cluster={cluster}
                    ownerReferences={resource.metadata?.ownerReferences}
                    namespace={resource.metadata?.namespace}
                  />
                </DescriptionListDescription>
              </DescriptionListGroup>
            </DescriptionList>
          </Stack>
        </PageSection>
        {resource.status?.conditions && <ResourceConditions conditions={resource.status.conditions} />}
      </PageSection>
    )
  }
  return <Fragment />
}
