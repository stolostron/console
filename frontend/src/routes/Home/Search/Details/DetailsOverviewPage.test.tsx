/* Copyright Contributors to the Open Cluster Management project */
// Copyright (c) 2022 Red Hat, Inc.

import { render, screen, waitFor } from '@testing-library/react'
import { createBrowserHistory } from 'history'
import { Router } from 'react-router-dom'
import { RecoilRoot } from 'recoil'
import { nockIgnoreApiPaths, nockIgnoreRBAC } from '../../../../lib/nock-util'
import DetailsOverviewPage, {
  LablesGroup,
  OwnerReferences,
  ResourceConditions,
  ResourceSearchLink,
} from './DetailsOverviewPage'

describe('DetailsOverviewPage', () => {
  beforeEach(async () => {
    nockIgnoreRBAC()
    nockIgnoreApiPaths()
  })

  it('Should correctly return ResourceSearchLink', async () => {
    render(
      <RecoilRoot>
        <Router history={createBrowserHistory()}>
          <ResourceSearchLink
            cluster={'test-cluster'}
            apiversion={'v1/beta1'}
            kind={'test-kind'}
            name={'test-name'}
            namespace={'test-namespace'}
          />
        </Router>
      </RecoilRoot>
    )

    // Test that the component has rendered correctly with data
    await waitFor(() => expect(screen.queryByText('test-name')).toBeTruthy())
  })

  it('Should correctly return empty LablesGroup', async () => {
    render(
      <RecoilRoot>
        <Router history={createBrowserHistory()}>
          <LablesGroup labels={{}} />
        </Router>
      </RecoilRoot>
    )

    // Test that the component has rendered correctly with data
    await waitFor(() => expect(screen.queryByText('No labels')).toBeTruthy())
  })

  it('Should correctly return LablesGroup', async () => {
    render(
      <RecoilRoot>
        <Router history={createBrowserHistory()}>
          <LablesGroup labels={{ test: 'test1', region: 'east' }} />
        </Router>
      </RecoilRoot>
    )

    // Test that the component has rendered correctly with data
    await waitFor(() => expect(screen.queryByText('test=test1')).toBeTruthy())
    await waitFor(() => expect(screen.queryByText('region=east')).toBeTruthy())
  })

  it('Should correctly return empty OwnerReferences', async () => {
    render(
      <RecoilRoot>
        <Router history={createBrowserHistory()}>
          <OwnerReferences namespace={'test-ns'} cluster={'test-cluster'} />
        </Router>
      </RecoilRoot>
    )

    // Test that the component has rendered correctly with data
    await waitFor(() => expect(screen.queryByText('No owners')).toBeTruthy())
  })

  it('Should correctly return OwnerReferences', async () => {
    render(
      <RecoilRoot>
        <Router history={createBrowserHistory()}>
          <OwnerReferences
            ownerReferences={[
              {
                apiVersion: 'v1/alpha1',
                kind: 'kind',
                name: 'test-name',
              },
            ]}
            namespace={'test-ns'}
            cluster={'test-cluster'}
          />
        </Router>
      </RecoilRoot>
    )

    // Test that the component has rendered correctly with data
    await waitFor(() => expect(screen.queryByText('test-name')).toBeTruthy())
  })

  it('Should correctly return empty ResourceConditions', async () => {
    render(
      <RecoilRoot>
        <Router history={createBrowserHistory()}>
          <ResourceConditions conditions={[]} />
        </Router>
      </RecoilRoot>
    )

    // Test that the component has rendered correctly with data
    await waitFor(() => expect(screen.queryByText('No conditions found')).toBeTruthy())
  })

  it('Should correctly return ResourceConditions', async () => {
    render(
      <RecoilRoot>
        <Router history={createBrowserHistory()}>
          <ResourceConditions
            conditions={[
              {
                type: 'Initialized',
                status: 'True',
              },
              {
                type: 'Available',
                status: 'True',
                lastTransitionTime: '2022-10-31T09:26:14Z',
                reason: 'MinimumReplicasAvailable',
                message: 'Deployment has minimum availability.',
              },
            ]}
          />
        </Router>
      </RecoilRoot>
    )

    // Test that the component has rendered correctly with data
    await waitFor(() => expect(screen.queryByText('Initialized')).toBeTruthy())
    await waitFor(() => expect(screen.queryByText('Available')).toBeTruthy())
    await waitFor(() => expect(screen.queryByText('MinimumReplicasAvailable')).toBeTruthy())
    await waitFor(() => expect(screen.queryByText('Deployment has minimum availability.')).toBeTruthy())
  })

  // Full page tests
  it('Should correctly return DetailsOverviewPage in loading state', async () => {
    render(
      <RecoilRoot>
        <Router history={createBrowserHistory()}>
          <DetailsOverviewPage
            cluster={'test-cluster'}
            resource={{
              apiVersion: 'v1',
              kind: 'Pod',
            }}
            loading={true}
            error={''}
          />
        </Router>
      </RecoilRoot>
    )

    // Test that the component has rendered correctly with data
    await waitFor(() => expect(screen.queryByText('Loading')).toBeTruthy())
  })

  it('Should correctly return DetailsOverviewPage with errors', async () => {
    render(
      <RecoilRoot>
        <Router history={createBrowserHistory()}>
          <DetailsOverviewPage
            cluster={'test-cluster'}
            resource={{
              apiVersion: 'v1',
              kind: 'Pod',
            }}
            loading={false}
            error={'Error getting resource'}
          />
        </Router>
      </RecoilRoot>
    )

    // Test that the component has rendered correctly with data
    await waitFor(() => expect(screen.queryByText('Error querying for resource:')).toBeTruthy())
  })

  it('Should correctly return DetailsOverviewPage', async () => {
    render(
      <RecoilRoot>
        <Router history={createBrowserHistory()}>
          <DetailsOverviewPage
            cluster={'local-cluster'}
            resource={
              {
                kind: 'Deployment',
                apiVersion: 'apps/v1',
                metadata: {
                  name: 'application-manager',
                  namespace: 'open-cluster-management-agent-addon',
                  uid: '4185c145-eb24-472b-b35b-92fdeaaa1b6a',
                  resourceVersion: '105557',
                  creationTimestamp: '2022-10-31T11:24:07Z',
                  labels: { component: 'application-manager' },
                  annotations: { 'deployment.kubernetes.io/revision': '1' },
                  ownerReferences: [
                    {
                      apiVersion: 'work.open-cluster-management.io/v1',
                      kind: 'AppliedManifestWork',
                      name: 'b87371581c40be6a58d412e4f67b1a024e1520f3e0b23ca80d5b30e5a1faa960-addon-application-manager-deploy',
                      uid: '89a634bb-8106-44c6-bd02-87c847e11fa9',
                    },
                  ],
                },
                spec: {
                  selector: {
                    matchLabels: {
                      'app.kubernetes.io/component': 'exporter',
                      'app.kubernetes.io/name': 'kube-state-metrics',
                      'app.kubernetes.io/part-of': 'openshift-monitoring',
                    },
                  },
                  template: {
                    spec: {
                      nodeSelector: {
                        'kubernetes.io/os': 'linux',
                      },
                    },
                  },
                },
                status: {
                  observedGeneration: 1,
                  replicas: 1,
                  updatedReplicas: 1,
                  readyReplicas: 1,
                  availableReplicas: 1,
                  conditions: [
                    {
                      type: 'Progressing',
                      status: 'True',
                      lastUpdateTime: '2022-10-31T11:24:37Z',
                      lastTransitionTime: '2022-10-31T11:24:07Z',
                      reason: 'NewReplicaSetAvailable',
                      message: 'ReplicaSet "application-manager-84c48b9597" has successfully progressed.',
                    },
                    {
                      type: 'Available',
                      status: 'True',
                      lastUpdateTime: '2022-10-31T11:34:24Z',
                      lastTransitionTime: '2022-10-31T11:34:24Z',
                      reason: 'MinimumReplicasAvailable',
                      message: 'Deployment has minimum availability.',
                    },
                  ],
                },
              } as any
            }
            loading={false}
            error={''}
          />
        </Router>
      </RecoilRoot>
    )

    // Test that the component has rendered correctly with data
    await waitFor(() => expect(screen.queryByText('application-manager')).toBeTruthy())
    await waitFor(() => expect(screen.queryByText('component=application-manager')).toBeTruthy())
    await waitFor(() => expect(screen.queryByText('kubernetes.io/os=linux')).toBeTruthy())
    await waitFor(() =>
      expect(
        screen.queryByText(
          'app.kubernetes.io/component=exporter, app.kubernetes.io/name=kube-state-metrics, app.kubernetes.io/part-of=openshift-monitoring'
        )
      ).toBeTruthy()
    )
    await waitFor(() => expect(screen.queryByText('Progressing')).toBeTruthy())
    await waitFor(() => expect(screen.queryByText('NewReplicaSetAvailable')).toBeTruthy())
    await waitFor(() =>
      expect(
        screen.queryByText('ReplicaSet "application-manager-84c48b9597" has successfully progressed.')
      ).toBeTruthy()
    )
  })
})
