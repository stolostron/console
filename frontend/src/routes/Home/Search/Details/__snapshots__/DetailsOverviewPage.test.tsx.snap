// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailsOverviewPage Should correctly return empty date 1`] = `"-"`;

exports[`DetailsOverviewPage Should correctly return getDate 1`] = `
<div
  style={
    Object {
      "alignItems": "baseline",
      "display": "flex",
    }
  }
>
  <GlobeAmericasIcon
    color="currentColor"
    noVerticalAlign={false}
    size="sm"
    style={
      Object {
        "marginRight": "5px",
      }
    }
  />
  <Tooltip
    content={
      Array [
        <span
          className="co-nowrap"
        >
          Oct 31, 2022, 11:34 AM UTC
        </span>,
      ]
    }
  >
    <span
      data-test="timestamp"
    >
      Oct 31, 2022, 7:34 AM
    </span>
  </Tooltip>
</div>
`;
